File: src/models/qrCodeSession.js
/**
 * 扫码登录会话管理 - 异步 I/O 版本
 */
const crypto = require('crypto');
const fs = require('fs').promises;
const path = require('path');

// 二维码有效期 (5分钟)
const QRCODE_EXPIRY = 5 * 60 * 1000;

// 数据文件路径
const QR_FILE = path.join(__dirname, '../../data/qrSessions.json');

// 状态枚举
const QRStatus = {
  PENDING: 'pending',
  SCANNED: 'scanned',
  CONFIRMED: 'confirmed',
  CANCELLED: 'cancelled',
  EXPIRED: 'expired'
};

// 内存缓存
let qrCache = null;
let cacheTime = 0;
const CACHE_TTL = 2000; // 2秒缓存

/**
 * 确保数据目录存在
 */
async function ensureDataDir() {
  const dataDir = path.dirname(QR_FILE);
  await fs.mkdir(dataDir, { recursive: true }).catch(err => {
    if (err.code !== 'EEXIST') throw err;
  });
}

/**
 * 读取所有二维码会话（带缓存）
 */
async function readQRSessions() {
  // 缓存命中
  if (qrCache !== null && (Date.now() - cacheTime) < CACHE_TTL) {
    return qrCache;
  }
  
  try {
    const data = await fs.readFile(QR_FILE, 'utf8');
    qrCache = JSON.parse(data);
    cacheTime = Date.now();
    return qrCache;
  } catch (err) {
    if (err.code === 'ENOENT') {
      qrCache = {};
      cacheTime = Date.now();
      return qrCache;
    }
    throw err;
  }
}

/**
 * 保存所有二维码会话
 */
async function writeQRSessions(sessions) {
  await ensureDataDir();
  await fs.writeFile(QR_FILE, JSON.stringify(sessions, null, 2), 'utf8');
  // 更新缓存
  qrCache = sessions;
  cacheTime = Date.now();
}

/**
 * 创建二维码会话
 * @returns {Promise<object>} { token, expiresAt }
 */
async function createQRSession() {
  const sessions = await readQRSessions();
  const token = crypto.randomBytes(32).toString('hex');
  
  const qrSession = {
    token,
    status: QRStatus.PENDING,
    createdAt: Date.now(),
    expiresAt: Date.now() + QRCODE_EXPIRY,
    sessionId: null,
    username: null
  };
  
  sessions[token] = qrSession;
  await writeQRSessions(sessions);
  
  return { token, expiresAt: qrSession.expiresAt };
}

/**
 * 获取二维码会话
 * @param {string} token 
 * @returns {Promise<object|null>}
 */
async function getQRSession(token) {
  if (!token) return null;
  
  const sessions = await readQRSessions();
  const qrSession = sessions[token];
  
  if (!qrSession) return null;
  
  // 检查是否过期
  if (Date.now() > qrSession.expiresAt) {
    qrSession.status = QRStatus.EXPIRED;
    sessions[token] = qrSession;
    await writeQRSessions(sessions);
  }
  
  return qrSession;
}

/**
 * 标记为已扫描
 * @param {string} token 
 * @returns {Promise<boolean>}
 */
async function markScanned(token) {
  const sessions = await readQRSessions();
  const qrSession = sessions[token];
  
  if (!qrSession || qrSession.status !== QRStatus.PENDING) {
    return false;
  }
  
  qrSession.status = QRStatus.SCANNED;
  sessions[token] = qrSession;
  await writeQRSessions(sessions);
  
  return true;
}

/**
 * 确认登录
 * @param {string} token 
 * @param {string} sessionId 
 * @param {string} username 
 * @returns {Promise<boolean>}
 */
async function confirmLogin(token, sessionId, username) {
  const sessions = await readQRSessions();
  const qrSession = sessions[token];
  
  if (!qrSession || qrSession.status !== QRStatus.SCANNED) {
    return false;
  }
  
  qrSession.status = QRStatus.CONFIRMED;
  qrSession.sessionId = sessionId;
  qrSession.username = username;
  sessions[token] = qrSession;
  await writeQRSessions(sessions);
  
  return true;
}

/**
 * 取消登录
 * @param {string} token 
 */
async function cancelLogin(token) {
  const sessions = await readQRSessions();
  const qrSession = sessions[token];
  
  if (qrSession && qrSession.status === QRStatus.SCANNED) {
    qrSession.status = QRStatus.CANCELLED;
    sessions[token] = qrSession;
    await writeQRSessions(sessions);
  }
}

/**
 * 删除二维码会话
 * @param {string} token 
 */
async function deleteQRSession(token) {
  const sessions = await readQRSessions();
  if (sessions[token]) {
    delete sessions[token];
    await writeQRSessions(sessions);
  }
}

/**
 * 清理过期会话
 */
async function cleanExpiredQRSessions() {
  const sessions = await readQRSessions();
  const now = Date.now();
  let changed = false;
  
  for (const [token, session] of Object.entries(sessions)) {
    if (now > session.expiresAt) {
      delete sessions[token];
      changed = true;
    }
  }
  
  if (changed) {
    await writeQRSessions(sessions);
  }
}

/**
 * 清除缓存
 */
function clearCache() {
  qrCache = null;
  cacheTime = 0;
}

// 每5分钟清理过期会话
setInterval(cleanExpiredQRSessions, 5 * 60 * 1000);

module.exports = {
  createQRSession,
  getQRSession,
  markScanned,
  confirmLogin,
  cancelLogin,
  deleteQRSession,
  cleanExpiredQRSessions,
  clearCache,
  QRStatus
};

File: go-impl/internal/model/qrCodeSession.go
package model

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// QRStatus 二维码状态
type QRStatus string

const (
	QRStatusPending   QRStatus = "pending"
	QRStatusScanned   QRStatus = "scanned"
	QRStatusConfirmed QRStatus = "confirmed"
	QRStatusCancelled QRStatus = "cancelled"
	QRStatusExpired   QRStatus = "expired"
)

// QRSession 二维码会话
type QRSession struct {
	Token     string   `json:"token"`
	Status    QRStatus `json:"status"`
	CreatedAt int64    `json:"createdAt"`
	ExpiresAt int64    `json:"expiresAt"`
	SessionID string   `json:"sessionId"`
	Username  string   `json:"username"`
}

// QRCodeManager 二维码会话管理器
type QRCodeManager struct {
	filePath string
	sessions map[string]*QRSession
	mu       sync.RWMutex
	expiry   time.Duration
}

var (
	qrCodeManager *QRCodeManager
	qrOnce        sync.Once
)

// GetQRCodeManager 获取二维码管理器单例
func GetQRCodeManager() *QRCodeManager {
	qrOnce.Do(func() {
		execPath, _ := os.Executable()
		baseDir := filepath.Dir(execPath)
		qrCodeManager = &QRCodeManager{
			filePath: filepath.Join(baseDir, "data", "qrSessions.json"),
			sessions: make(map[string]*QRSession),
			expiry:   5 * time.Minute,
		}
		qrCodeManager.load()
		go qrCodeManager.startCleanup()
	})
	return qrCodeManager
}

func (qm *QRCodeManager) load() {
	data, err := os.ReadFile(qm.filePath)
	if err != nil {
		return
	}
	json.Unmarshal(data, &qm.sessions)
}

func (qm *QRCodeManager) save() {
	os.MkdirAll(filepath.Dir(qm.filePath), 0755)
	data, _ := json.MarshalIndent(qm.sessions, "", "  ")
	os.WriteFile(qm.filePath, data, 0644)
}

// CreateQRSession 创建二维码会话
func (qm *QRCodeManager) CreateQRSession() (string, int64) {
	qm.mu.Lock()
	defer qm.mu.Unlock()
	
	bytes := make([]byte, 32)
	rand.Read(bytes)
	token := hex.EncodeToString(bytes)
	
	now := time.Now()
	session := &QRSession{
		Token:     token,
		Status:    QRStatusPending,
		CreatedAt: now.UnixMilli(),
		ExpiresAt: now.Add(qm.expiry).UnixMilli(),
	}
	
	qm.sessions[token] = session
	qm.save()
	
	return token, session.ExpiresAt
}

// GetQRSession 获取二维码会话
func (qm *QRCodeManager) GetQRSession(token string) *QRSession {
	qm.mu.RLock()
	defer qm.mu.RUnlock()
	
	session, ok := qm.sessions[token]
	if !ok {
		return nil
	}
	
	// 检查过期
	if time.Now().UnixMilli() > session.ExpiresAt {
		session.Status = QRStatusExpired
	}
	
	return session
}

// MarkScanned 标记为已扫描
func (qm *QRCodeManager) MarkScanned(token string) bool {
	qm.mu.Lock()
	defer qm.mu.Unlock()
	
	session, ok := qm.sessions[token]
	if !ok || session.Status != QRStatusPending {
		return false
	}
	
	session.Status = QRStatusScanned
	qm.save()
	return true
}

// ConfirmLogin 确认登录
func (qm *QRCodeManager) ConfirmLogin(token, sessionID, username string) bool {
	qm.mu.Lock()
	defer qm.mu.Unlock()
	
	session, ok := qm.sessions[token]
	if !ok || session.Status != QRStatusScanned {
		return false
	}
	
	session.Status = QRStatusConfirmed
	session.SessionID = sessionID
	session.Username = username
	qm.save()
	return true
}

// CancelLogin 取消登录
func (qm *QRCodeManager) CancelLogin(token string) {
	qm.mu.Lock()
	defer qm.mu.Unlock()
	
	session, ok := qm.sessions[token]
	if ok && session.Status == QRStatusScanned {
		session.Status = QRStatusCancelled
		qm.save()
	}
}

// DeleteQRSession 删除二维码会话
func (qm *QRCodeManager) DeleteQRSession(token string) {
	qm.mu.Lock()
	defer qm.mu.Unlock()
	
	delete(qm.sessions, token)
	qm.save()
}

func (qm *QRCodeManager) startCleanup() {
	ticker := time.NewTicker(5 * time.Minute)
	for range ticker.C {
		qm.cleanup()
	}
}

func (qm *QRCodeManager) cleanup() {
	qm.mu.Lock()
	defer qm.mu.Unlock()
	
	now := time.Now().UnixMilli()
	for token, session := range qm.sessions {
		if now > session.ExpiresAt {
			delete(qm.sessions, token)
		}
	}
	qm.save()
}
